### DISCUSSION TITLE
Using PLV8 for RPC Calls with PostgREST

### DISCUSSION THREAD

**Discussion by robertsosinski** [PostgREST/postgrest#2084] [Q&A] [answered]
Hello y'all!

Looking into making a few RPC functions with PLV8 and having a performance question.  These functions will receive significant web traffic after being made available via PostgREST, and the specific question I have is around this part in the docs at https://plv8.github.io

> In PLV8, each session has one global JS runtime context. This enables function invocations at low cost, and sharing common object among the functions. However, for the security reasons, if the user switches to another with SET ROLE command, a new JS runtime context is initialized and used separately. This prevents the risk of unexpected information leaking.

I know that PostgREST extensively uses the SET ROLE feature to locally scope privileges for each request made. As such, would each request initialize a separate JS runtime?

In more general terms, would anyone see any problems with making a high traffic RPC function with PLV8 instead of PLSQL via PostgREST?

Appreciate any help, and appreciate as always the great work on PostgREST!

**Comment by wolfgangwalther** [comment]
I have not worked with plv8, yet, but I imagine you could reduce the number of effective `SET ROLE` comands a lot by using [shared roles](https://postgrest.org/en/v9.0/auth.html#web-users-sharing-role). In theory you could even run the authenticated requests on the same role as the anonymous role and do all privilege checking only via RLS or inside the RPC... although you'd need to be careful with privileges in that case.

This would not remove the `SET ROLE` calls, but it would basically lead to issuing the same `SET ROLE x` call again and again within the session that is kept open. Now, I don't know whether plv8 creates a new context whenever the `SET ROLE` command is issued, independently of the role is actually changed - or only when the role is effectively switched to another role. It would only improve things in the latter case, obviously.

But before going anywhere in that direction, I'd just test the performance for real. There is no point in trying to optimize something that might not even be a problem.

**Comment by wolfgangwalther** [comment, marked as answer]
Did you try using plv8 and have some experience to share by now?
